saleor/graphql/warehouse/dataloaders.py

```diff
@@ -0,0 +1,88 @@
+from collections import defaultdict
+from typing import DefaultDict, Iterable, List, Optional, Tuple
+
+from django.conf import settings
+from django.db.models import Q, Sum
+from django.db.models.functions import Coalesce
+
+from ...warehouse.models import Stock
+from ..core.dataloaders import DataLoader
+
+CountryCode = Optional[str]
+VariantIdAndCountryCode = Tuple[int, CountryCode]
+
+
+class AvailableQuantityByProductVariantIdAndCountryCodeLoader(
+    DataLoader[VariantIdAndCountryCode, int]
+):
+    """Calculates available variant quantity based on variant ID and country code.
+
+    For each country code, for each shipping zone supporting that country,
+    calculate the maximum available quantity, then return either that number
+    or the maximum allowed checkout quantity, whichever is lower.
+    """
+
+    context_key = "stock_by_productvariant_and_country"
+
+    def batch_load(self, keys):
+        # Split the list of keys by country first. A typical query will only touch
+        # a handful of unique countries but may access thousands of product variants
+        # so it's cheaper to execute one query per country.
+        variants_by_country: DefaultDict[CountryCode, List[int]] = defaultdict(list)
+        for variant_id, country_code in keys:
+            variants_by_country[country_code].append(variant_id)
+
+        # For each country code execute a single query for all product variants.
+        quantity_by_variant_and_country: DefaultDict[
+            VariantIdAndCountryCode, int
+        ] = defaultdict(int)
+        for country_code, variant_ids in variants_by_country.items():
+            quantities = self.batch_load_country(country_code, variant_ids)
+            for variant_id, quantity in quantities:
+                quantity_by_variant_and_country[(variant_id, country_code)] = quantity
+
+        return [quantity_by_variant_and_country[key] for key in keys]
+
+    def batch_load_country(
+        self, country_code: CountryCode, variant_ids: Iterable[int]
+    ) -> Iterable[Tuple[int, int]]:
+        filters = Q(product_variant_id__in=variant_ids)
+        if country_code:
+            filters &= Q(warehouse__shipping_zones__countries__contains=country_code)
+        results = (
+            Stock.objects.filter(filters)
+            .annotate(
+                available_quantity=Sum("quantity", distinct=True)
```

Comment:
Why `Sum` aggregation on `quantity`, which is a single value in each stock row? Wouldn't `F("quantity")` be enough? In `StockQuerySet` queryset we the following function:

```
def annotate_available_quantity(self):
        return self.annotate(
            available_quantity=F("quantity")
            - Coalesce(Sum("allocations__quantity_allocated"), 0)
        )
```

Ideally, we should reuse this function in all places that need to calculate the available quantity on stock records (there are few more functions that do that in the `warehouse/availability.py` module.